### PR TITLE
fix: Frontend doesn't handle left semi join correctly.

### PR DESCRIPTION
**Pull request by jon-chuang**
## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***
Changed `prune_cols` in `LogicalJoin` to take into account semi/anti join for `required_cols`
## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
Fixes: #1849

### CODE REVIEW

**Comment by TennyZhuang** [comment]
Please sign your CLA, thanks!

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 1]
```diff
@@ -281,35 +289,55 @@ impl ColPrunable for LogicalJoin {
...
 
-        if required_cols == &left_right_required_cols {
+        let required_input_cols_in_output = if self.is_left_join() {
+            left_input_cols
+        } else if self.is_right_join() {
+            right_input_cols
+        } else {
+            left_right_input_cols.clone()
+        };
+        if required_cols == &required_input_cols_in_output {
```
Comment (jon-chuang) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:337]: Can remove clone

**Comment by st1page** [comment]
The fix LGTM, but the main branch version you derived from is a little old. could you please update the branch and fix the LeftSemi join's behavior of `out_column_num`, `o2l_col_mapping_inner` and `o2r_col_mapping_inner`?

**Comment by likg227** [comment]
Could you please remove the FIXME and fix those three functions mentioned in the issue? After that you can run planner test to test whether this pr is correct.

**Comment by jon-chuang** [comment]
done :)

**Review by likg227 — APPROVED** [review]
LGTM. Thx for your contribution!

**Review by st1page — APPROVED** [review]
LGTM, Thanks for your contribution!

**Comment by TennyZhuang** [comment]
https://github.com/singularity-data/risingwave/runs/6345507825?check_suite_focus=true CI failed, please fix the clippy check

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 2]
```diff
@@ -508,6 +523,80 @@ mod tests {
...
+            let mut required_cols = FixedBitSet::with_capacity(3);
+            // key 0 is never used in the join (always key 1)
+            required_cols.extend(vec![0]);
+            // should not panic here
+            let plan = join.prune_col(&required_cols);
+            // Check that the join has been wrapped in a projection
+            let as_plan = plan.as_logical_project().unwrap();
+            // Check the result
+            assert_eq!(as_plan.schema().fields().len(), 1);
+            assert_eq!(as_plan.schema().fields()[0], fields[offset + 0]);
```
Comment (TennyZhuang) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:583]: clippy failed on `offset + 0`. If you are sure it's more readable, you can add allow(clippy::?) attribute similar to https://stackoverflow.com/questions/55402812/how-to-disable-a-clippy-lint-for-a-single-line-block

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 3]
```diff
@@ -508,6 +523,80 @@ mod tests {
...
+            // Perform the prune
+            let mut required_cols = FixedBitSet::with_capacity(3);
+            required_cols.extend(vec![0, 1, 2]);
+            // should not panic here
+            let plan = join.prune_col(&required_cols);
+            // Check that the join has not been wrapped in a projection
+            let as_plan = plan.as_logical_join().unwrap();
+            // Check the result
+            assert_eq!(as_plan.schema().fields().len(), 3);
+            assert_eq!(as_plan.schema().fields()[0], fields[offset + 0]);
```
Comment (TennyZhuang) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:594]: ditto

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 4]
```diff
@@ -293,35 +290,53 @@ impl ColPrunable for LogicalJoin {
...
+                resized_required_cols.insert(left_len + i);
+            } else {
+                resized_required_cols.insert(i);
+            }
+        });
+
+        let mut visitor = CollectInputRef::new(resized_required_cols.clone());
         self.on.visit_expr(&mut visitor);
-        let left_right_required_cols = visitor.collect();
+        let left_right_input_cols = visitor.collect();
```
Comment (xxchan) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:306]: Could you explain why did you change the variables' naming?

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 5]
```diff
@@ -508,6 +523,80 @@ mod tests {
         assert_eq!(right.schema().fields(), &fields[3..4]);
     }
 
+    /// Semi join panicked previously at `prune_col`. Add test to prevent regression.
+    #[tokio::test]
+    async fn test_semi_join_does_not_panic() {
```
Comment (xxchan) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:528]: What about
```suggestion
    async fn test_prune_semi_join() {
```

**Review by xxchan — APPROVED** [review]
LGTM! Please resolve remaining comments.

File: src/frontend/src/optimizer/plan_node/logical_join.rs (hunk 4 above)
Comment (jon-chuang) [reply src/frontend/src/optimizer/plan_node/logical_join.rs]: I thought it was more readable while coding. But I think its fine now that I read again.

**Comment by xxchan** [comment]
You can run `./risedev do-apply-planner-test` to update the plan test and check whether the new one looks good.

**Comment by jon-chuang** [comment]
I think the sql planner tests are failing as previously, semijoin did not have the correct output. Should I change those tests?

**Comment by StrikeW** [comment]
merge?
